File: src/main/java/com/example/restapi/service/UserService.java
package com.example.restapi.service;

import com.example.restapi.model.User;
import com.example.restapi.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class UserService {
    
    private final UserRepository userRepository;
    
    @Autowired
    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }
    
    /**
     * Get all users
     * @return List of all users
     */
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }
    
    /**
     * Get user by ID
     * @param id the user ID
     * @return Optional containing the user if found
     */
    public Optional<User> getUserById(Long id) {
        return userRepository.findById(id);
    }
    
    /**
     * Create a new user
     * @param user the user to create
     * @return the created user
     * @throws IllegalArgumentException if email already exists
     */
    public User createUser(User user) {
        if (userRepository.existsByEmail(user.getEmail())) {
            throw new IllegalArgumentException("User with email " + user.getEmail() + " already exists");
        }
        return userRepository.save(user);
    }
    
    /**
     * Update an existing user
     * @param id the user ID
     * @param userDetails the updated user details
     * @return the updated user
     * @throws IllegalArgumentException if user not found or email already exists
     */
    public User updateUser(Long id, User userDetails) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("User not found with id: " + id));
        
        // Check if email is being changed and if it already exists
        if (!user.getEmail().equals(userDetails.getEmail()) && 
            userRepository.existsByEmail(userDetails.getEmail())) {
            throw new IllegalArgumentException("User with email " + userDetails.getEmail() + " already exists");
        }
        
        user.setName(userDetails.getName());
        user.setEmail(userDetails.getEmail());
        user.setDescription(userDetails.getDescription());
        
        return userRepository.save(user);
    }
    
    /**
     * Delete a user
     * @param id the user ID
     * @throws IllegalArgumentException if user not found
     */
    public void deleteUser(Long id) {
        if (!userRepository.existsById(id)) {
            throw new IllegalArgumentException("User not found with id: " + id);
        }
        userRepository.deleteById(id);
    }
    
    /**
     * Get user by email
     * @param email the email to search for
     * @return Optional containing the user if found
     */
    public Optional<User> getUserByEmail(String email) {
        return userRepository.findByEmail(email);
    }
}

File: advanced-level/java-to-python-modernization/src/api/services/user_service.py
"""User service for business logic."""
from sqlalchemy.orm import Session
from typing import Optional, List
from fastapi import HTTPException

from ..repositories.user_repository import UserRepository, User
from ..models.user import UserCreate, UserUpdate


class UserService:
    """Service for user business logic."""
    
    def __init__(self, db: Session):
        """
        Initialize service with database session.
        
        Args:
            db: SQLAlchemy database session
        """
        self.repository = UserRepository(db)
    
    def get_all_users(self) -> List[User]:
        """
        Get all users.
        
        Returns:
            List of all User objects
        """
        return self.repository.get_all()
    
    def get_user_by_id(self, user_id: int) -> Optional[User]:
        """
        Get user by ID.
        
        Args:
            user_id: User ID to search for
            
        Returns:
            User object if found, None otherwise
        """
        return self.repository.get_by_id(user_id)
    
    def get_user_by_email(self, email: str) -> Optional[User]:
        """
        Get user by email address.
        
        Args:
            email: Email address to search for
            
        Returns:
            User object if found, None otherwise
        """
        return self.repository.get_by_email(email)
    
    def create_user(self, user_data: UserCreate) -> User:
        """
        Create a new user.
        
        Args:
            user_data: UserCreate model with user data
            
        Returns:
            Created User object
            
        Raises:
            HTTPException: If email already exists (400)
        """
        if self.repository.exists_by_email(user_data.email):
            raise HTTPException(
                status_code=400,
                detail=f"User with email {user_data.email} already exists"
            )
        
        user_dict = {
            "name": user_data.name,
            "email": user_data.email,
            "description": user_data.description
        }
        
        return self.repository.create(user_dict)
    
    def update_user(self, user_id: int, user_data: UserUpdate) -> User:
        """
        Update an existing user.
        
        Args:
            user_id: User ID to update
            user_data: UserUpdate model with updated data
            
        Returns:
            Updated User object
            
        Raises:
            HTTPException: If user not found (404) or email conflict (400)
        """
        user = self.repository.get_by_id(user_id)
        if not user:
            raise HTTPException(
                status_code=404,
                detail=f"User not found with id: {user_id}"
            )
        
        # Check email uniqueness if email is being updated
        if user_data.email is not None and user_data.email != user.email:
            if self.repository.exists_by_email(user_data.email):
                raise HTTPException(
                    status_code=400,
                    detail=f"User with email {user_data.email} already exists"
                )
        
        # Build update dictionary (only include non-None values)
        update_dict = {}
        if user_data.name is not None:
            update_dict["name"] = user_data.name
        if user_data.email is not None:
            update_dict["email"] = user_data.email
        if user_data.description is not None:
            update_dict["description"] = user_data.description
        
        updated_user = self.repository.update(user_id, update_dict)
        if not updated_user:
            raise HTTPException(
                status_code=404,
                detail=f"User not found with id: {user_id}"
            )
        
        return updated_user
    
    def delete_user(self, user_id: int) -> None:
        """
        Delete a user by ID.
        
        Args:
            user_id: User ID to delete
            
        Raises:
            HTTPException: If user not found (404)
        """
        if not self.repository.exists_by_id(user_id):
            raise HTTPException(
                status_code=404,
                detail=f"User not found with id: {user_id}"
            )
        
        self.repository.delete(user_id)

